### DISCUSSION TITLE
Attributes on READ_ONLY LDAP users cannot be saved

### DISCUSSION THREAD

**Discussion by LarsNeidhold** [keycloak/keycloak#24679] [Q&A] [answered]
Hi all,

I have a LDAP user federation, with edit mode set to READONLY, since I don´t want to write changes on username, email, first name, last name, and other mapped attributes back to the LDAP.

Now when I try to add a new attribute to a user of that user federation, I get a “Error updating group User is read only!”.

I´m a bit confused , since the documentation states for edit mode READONLY:

> You cannot change the username, email, first name, last name, and other mapped attributes. Keycloak shows an error anytime a user attempts to update these fields.

But the attribute I tried to add is definitely not a mapped attribute.

Is this a bug?

Regards
Lars

**Comment by MariiaAlekseichuk** [comment]
Hi.
I also wonder if it should work in new User Profile feature (declarative-user-profile)?

I want to add new attribute for LDAP user and make it writable. 
Herewith this new attribute is not mapped attribute and LDAP edit mode is READONLY.

**Comment by MariiaAlekseichuk** [comment]
Can anybody confirm if it's a bug or an expected behavior?

**Comment by schouffy** [comment]
I have the same issue.

**Comment by aeckl** [comment]
We also have this issue, is there any news on this?

**Comment by michalpukalo** [comment, marked as answer]
@aeckl @schouffy  @MariiaAlekseichuk  @LarsNeidhold  You need to make a change in your user federation settings. You have to select UNSYNCED instead of READ_ONLY this way it will be editable without syncing back changes to LDAP:
> READ_ONLY is a read-only LDAP store. WRITABLE means data will be synced back to LDAP on demand. UNSYNCED means user data will be imported, but not synced back to LDAP.


![image](https://github.com/user-attachments/assets/68bf07dc-1934-45d5-bc74-f65fad350b70)

**Reply by LarsNeidhold** [reply]
Thanks, this works.

Although I think the word _mapped_ should be removed from the documentation:

> READONLY
> You cannot change the username, email, first name, last name, and other **_mapped_** attributes. Keycloak shows an error anytime a user attempts to update these fields. Password updates are not supported.

Also, for those considering changing the _Edit Mode_: 

>When Keycloak creates the LDAP provider, Keycloak also creates a set of initial LDAP mappers. Keycloak configures these mappers based on a combination of the Vendor, Edit Mode, and Import Users switches. For example, when edit mode is UNSYNCED, Keycloak configures the mappers to read a particular user attribute from the database and not from the LDAP server. However, if you later change the edit mode, the mapper’s configuration does not change because it is impossible to detect if the configuration changes changed in UNSYNCED mode. Decide the Edit Mode when creating the LDAP provider. This note applies to Import Users switch also.

**Reply by oculos** [reply]
I totally agree. I also, because the way the documentation is written, thought that unmapped attribute could exist in Keycloak's database. 
This would actually be nice to have. The Unsynced mode requires - and correct me if I'm wrong - importing the users, which is something I don't want.

**Reply by LarsNeidhold** [reply]
> This would actually be nice to have. The Unsynced mode requires - and correct me if I'm wrong - importing the users, which is something I don't want.

@oculos: This is correct.

**Reply by michalpukalo** [reply]
Can't you like configure it as READ_ONLY without syncing and then change it to UNSYNCED? Wouldn't it work that way?

**Reply by oculos** [reply]
@michalpukalo It complains that you need to turn "Import users" on.
